RePORTER query scripts

Bankso opened the pull request:
This PR adds two scripts that can be used for collecting grant information from the [NIH RePORTER API](https://api.reporter.nih.gov/):
- reporter_project_query.py: queries the RePORTER API projects endpoint for grant information and returns the results as a CSV
- clean_reporter_results.py: takes the CSV from RePORTER, removes outdated and redundant entries, and returns a CSV

While the scripts were designed for the specific purpose of collecting subproject_id values, the input arguments and JSON payload content in reporter_project_query.py can be modified to capture other information available from the RePORTER API. The extractions and filtering criteria in clean_reporter_results.py can also be adjusted, if needed. As such, I imagine these scripts can be extended to capture most or all of the GrantView and ProjectView metadata we record.

Bankso committed:
Add script to query RePORTER

Initial commit of a script that creates and submits a JSON query for the RePORTER project endpoint, then converts results into a CSV.

Takes a space-separated list of grant numbers and output CSV path as input. Optionally takes a list of years.

Results includes the following project information: subproject_id, project_title, project_num, fiscal_year, project_end_date

Other available information that is possible to retrieve is documented here: https://api.reporter.nih.gov/#project-data-elements

Bankso committed:
Add clean_reporter_results

This script processes the CSV from reporter_project_query, specifically:
- extracts the main and year values for each grant number and adds them to the table
- removes entries not associated with the current or final year of a grant
- sorts entries by grant number
- returns a new CSV

Takes a CSV or list of CSVs to process and combine, with minimum columns subproject_id and project_num (in RePORTER format, e.g. 3U54CA234567-05)

Returns a CSV with the original columns, plus main_grant (e.g., CA234567) and grant_year (e.g., 05) added.

Filtering is achieved for the output CSV based on (grant number, grant year) combinations, where only the (number, year) pairs that have the highest year value for each grant are retained. Example:
Between (CA234567, 05) and (CA234567, 03), the 05 set would be retained in the filtered CSV

Bankso committed:
Update .gitignore

Don't sync outputs

Bankso committed:
Update clean_reporter_results.py

Reorg extract_for_filtering function to accept tuples as definitions for new column names, regex for extraction, and name of column to extract from

Bankso commented:
@aclayton555 @aditigopalan @vpchung This PR isn't urgent - I just forgot to request a review when I put it together.

vpchung commented on utils/reporter_project_query.py:
```diff
@@ -0,0 +1,158 @@
+
+    headers = {"content-type": "application/json"}
+
+    max_request_length = int(400)
+    grant_count = int(len(grants))
+    req_per_grant = int(20)
```
Because these are already integers, I'm curious for the reason behind the `int` type castings?

Bankso replied:
That's a good question! I don't 100% remember, but I think I added the `int` casting to address an error from the `/` operation on line 108, which was complaining about not being able to divide `str` objects

vpchung commented on utils/reporter_project_query.py:
```diff
@@ -0,0 +1,158 @@
+
+    for year in in_years:
+        year = int(year)
+        years.append(year)
```
This can be further "simplified" by using list comprehensions, e.g.

```python
grants = [g + "*" for g in in_grants]
years = [int(y) for y in in_years]
```

Bankso replied:
Good point, this has been updated!

vpchung commented on utils/reporter_project_query.py:
```diff
@@ -0,0 +1,158 @@
+        grant_chunks = [g.tolist() for g in np.array_split(grants, split_count)]
+
+    else:
+        grant_chunks = grants
```
I noticed that `grant_chunks` will either be a list of lists or it'll be a list of strings, depending on the number of grants to be queried --  I wanted to make sure this was intentional?

For example, for the `if` code chunk, I used the list of grant numbers from [grantview_synapse_storage_manifest_table](https://www.synapse.org/Synapse:syn53259587/tables/) and got a list of lists in return:

```python
>>> grants
['CA184897*', 'CA184898*', ..., 'CA284086*', 'CA284090*']
>>> split_count = math.ceil(grant_count / req_per_grant) + 1
>>> grant_chunks = [g.tolist() for g in np.array_split(grants, split_count)]
>>> grant_chunks
[['CA184897*', 'CA184898*', ...], ..., ['CA274504*', 'CA274506*',...]]
```

For the `else`, I used a smaller subset and got back a list of strings:

```python
>>> grant_chunks2 = grants_subset
>>> grant_chunks2
['CA184897*', 'CA184898*']
```

Bankso replied:
Thanks for pointing this out. I definitely wanted a list of lists in both cases, so I changed the else statement to `grant_chunks = [grants]`

vpchung commented on utils/reporter_project_query.py:
```diff
@@ -0,0 +1,158 @@
+            print(
+                f"\n\nNot quite! Your query: \n {query} \n is an invalid JSON string.\nPlease revise your query and try again."
+            )
+            break
```
I wonder if this check is needed?  Because `build_payload()` returns a JSON string, my assumption would be that `json.loads(query)` is always truth-y, therefore, is always valid? Unless there's a use-case I'm not considering.... (it _is_ Friday afternoon, after all 😎 )

Bankso replied:
You're right, this is not needed! I had that in there when I was evaluating ways to build JSON queries, but I moved the final `json_payload = json.dumps(payload)` into `build_payload` and neglected to remove the `json.loads` call in `main`

vpchung replied:
Ah, makes sense!

vpchung reviewed (approved):
Looks good to me!

Some minor comments.

Bankso committed:
Apply list comprehension to main inputs

Bankso committed:
Change description, remove redundant json check

Bankso committed:
Ensure build_payload gets list of lists

Since the JSON payload is built from a list of lists, adding an un-chunked list of grants to an otherwise empty list provides the correct input type

Bankso committed:
Address read_json warning

From pd.read_json call: "FutureWarning: Passing literal json to 'read_json' is deprecated and will be removed in a future version. To read from a literal string, wrap it in a 'StringIO' object."

Adding "StringIO" casting resolved the warning message.

Bankso committed:
Update .gitignore

Bankso committed:
Merge branch 'main' into scripts-for-reporter

Bankso commented:
Thank you for your comments, @vpchung!
